rs/canister_sandbox/src/sandbox_manager.rs

```diff
@@ -318,6 +319,30 @@ impl SandboxManager {
         }
     }
 
+    pub fn open_wasm_via_file(
+        &self,
+        wasm_id: WasmId,
+        serialized_module_file: File,
+    ) -> HypervisorResult<(Arc<EmbedderCache>, Duration)> {
+        let mut guard = self.repr.lock().unwrap();
+        assert!(
+            !guard.caches.contains_key(&wasm_id),
+            "Failed to open wasm session {}: id is already in use",
+            wasm_id,
+        );
+        let deserialization_timer = Instant::now();
+        let instance_pre = self
+            .embedder
+            .read_file_and_pre_instantiate(serialized_module_file);
+        let cache = Arc::new(EmbedderCache::new(instance_pre.clone()));
```

Comment:
Could we deserialize directly into an `Arc` and avoid the `clone` here?

Reply:
Also, EmbedderCache already contains an Arc, so why are we wrapping it in another one?

Reply:
I think the clone is just needed here for the error case where we store the error, but also return it. We can probably modify it to only clone along the error path. I think the reason for wrapping the `EmbedderCache` in an `Arc` is the same: it only contains an `Arc` in the `Ok` case. In the `Err` case there might be a direct `String` or something.

But having said that, we should maybe change this `cache` map to have the values be just a `HypervisorResult<InstancePre<...>>` when you do a cheap clone out in the `Ok` case and maybe a more expensive clone in the `Err` case. It could simplify the code a bunch and in the error case we should just be serializing the data to send back anyway.